File: HS25_SoCo_group_040-a3/zvfs.py
import sys
import struct
from pathlib import Path
import climage
import time


class ZestFileSystem:
    # Constants
    MAX_FILES = 32
    MAX_FILENAME = 31
    FILE_ENTRY_SIZE = 64
    HEADER_SIZE = 64
    HEADER_FORMAT = "<8s B B H H H H H I I I I H 26s"
    FILE_ENTRY_FORMAT = "<32s I I B B H Q 12s"

    def __init__(self, filename: str):
        self.filename = Path(filename)

    #Helper methods
    def _read_header(self, fs):
        fs.seek(0)
        return list(struct.unpack(self.HEADER_FORMAT, fs.read(self.HEADER_SIZE)))

    def _write_header(self, fs, header_list):
        fs.seek(0)
        fs.write(struct.pack(self.HEADER_FORMAT, *header_list))

    def _find_file_entry(self, fs, filename):
        header = self._read_header(fs)
        table_offset = header[8]
        fs.seek(table_offset)
        for _ in range(self.MAX_FILES):
            entry_pos = fs.tell()
            entry_data = fs.read(self.FILE_ENTRY_SIZE)
            if not entry_data or entry_data == b'\x00' * self.FILE_ENTRY_SIZE:
                continue
            entry = struct.unpack(self.FILE_ENTRY_FORMAT, entry_data)
            name_str = entry[0].split(b'\x00')[0].decode(errors='ignore')
            if name_str == filename:
                return entry, entry_pos
        return None, None

    #Public methods
    @staticmethod
    def create_new_fs(name: str):
        name = Path(name)
        magic = b"ZVFSDSK1"
        version = 1
        flags = 0
        file_count = 0
        file_capacity = ZestFileSystem.MAX_FILES
        file_entry_size = ZestFileSystem.FILE_ENTRY_SIZE
        file_table_offset = ZestFileSystem.HEADER_SIZE
        data_start_offset = file_table_offset + file_capacity * file_entry_size
        next_free_offset = data_start_offset
        free_entry_offset = 0
        deleted_files = 0
        reserved2 = b"\x00" * 26
        header = struct.pack(
            ZestFileSystem.HEADER_FORMAT,
            magic, version, flags, 0, file_count, file_capacity,
            file_entry_size, 0, file_table_offset, data_start_offset,
            next_free_offset, free_entry_offset, deleted_files, reserved2
        )
        empty_entry = b"\x00" * ZestFileSystem.FILE_ENTRY_SIZE
        file_entries = empty_entry * file_capacity
        with name.open("wb") as f:
            f.write(header)
            f.write(file_entries)
        print(f"Created new filesystem '{name}' with capacity for {file_capacity} files.")


    @staticmethod
    def get_info_fs(file_system: str):
        file_system = Path(file_system)
        if not file_system.exists():
            print(f"File '{file_system}' does not exist!")
            sys.exit(1)
        with file_system.open("rb") as f:
            header_data = f.read(ZestFileSystem.HEADER_SIZE)
        fields = [
            "magic", "version", "flags", "reserved0", "file_count",
            "file_capacity", "file_entry_size", "reserved1",
            "file_table_offset", "data_start_offset", "next_free_offset",
            "free_entry_offset", "deleted_files", "reserved2"
        ]
        values = struct.unpack(ZestFileSystem.HEADER_FORMAT, header_data)
        header = dict(zip(fields, values))
        print(f"File system: {file_system}")
        print(f"Magic: {header['magic'].decode(errors='ignore')}")
        print(f"Version: {header['version']}")
        print(f"Files present: {header['file_count']}")
        print(f"Free entries: {header['file_capacity'] - header['file_count']}")
        print(f"Deleted files: {header['deleted_files']}")
        print(f"Total size: {file_system.stat().st_size} bytes")


    def add_file(self, fs_name: str, file_path: str):
        fs_name = Path(fs_name)
        file_path = Path(file_path)
        if not fs_name.exists():
            print(f"File system '{fs_name}' does not exist!")
            sys.exit(1)
        if not file_path.exists():
            print(f"Source file '{file_path}' does not exist!")
            sys.exit(1)
        data = file_path.read_bytes()
        filename = file_path.name
        if len(filename) > self.MAX_FILENAME:
            print(f"Filename '{filename}' too long (max {self.MAX_FILENAME})")
            sys.exit(1)
        with fs_name.open("r+b") as fs:
            header = self._read_header(fs)
            count, capacity, table_offset, next_free = header[4], header[5], header[8], header[10]
            if count >= capacity:
                print("ERROR: File system is full.")
                sys.exit(1)
            fs.seek(table_offset)
            free_idx = None
            for i in range(self.MAX_FILES):
                if fs.read(self.FILE_ENTRY_SIZE) == b"\x00" * self.FILE_ENTRY_SIZE:
                    free_idx = i
                    break
            if free_idx is None:
                print("ERROR: No free file entry found.")
                sys.exit(1)
            pad_len = (64 - (len(data) % 64)) % 64
            data_padded = data + b"\x00" * pad_len
            name_bytes = filename.encode('utf-8').ljust(32, b'\x00')
            start_offset = next_free
            file_size = len(data)
            entry = struct.pack(
                self.FILE_ENTRY_FORMAT, name_bytes, start_offset, file_size,
                0, 0, 0, int(time.time()), b"\x00" * 12
            )
            fs.seek(table_offset + free_idx * self.FILE_ENTRY_SIZE)
            fs.write(entry)
            fs.seek(start_offset)
            fs.write(data_padded)
            header[4] += 1
            header[10] += len(data_padded)
            self._write_header(fs, header)
        print(f"Added '{filename}' ({file_size} bytes, padded to {len(data_padded)} bytes) at offset {start_offset}.")


    def get_file(self, fs_name: str, filename: str):
        fs_name = Path(fs_name)
        out_path = Path.cwd() / filename
        if not fs_name.exists():
            print(f"File system '{fs_name}' does not exist!")
            sys.exit(1)
        with fs_name.open("rb") as fs:
            entry, _ = self._find_file_entry(fs, filename)
            if not entry:
                print(f"File '{filename}' not found in filesystem.")
                return
            start_offset, file_size, flag_field = entry[1], entry[2], entry[4]
            if flag_field == 1:
                print(f"File '{filename}' is deleted.")
                return
            fs.seek(start_offset)
            data = fs.read(file_size)
            out_path.write_bytes(data)
            print(f"Extracted '{filename}' to '{out_path}' ({file_size} bytes).")


    def cat_file(self, fs_name: str, filename: str):
        fs_name = Path(fs_name)
        if not fs_name.exists():
            print(f"File system '{fs_name}' does not exist!")
            sys.exit(1)
        with fs_name.open("rb") as fs:
            entry, _ = self._find_file_entry(fs, filename)
            if not entry:
                print(f"File '{filename}' not found in filesystem.")
                return
            start_offset, length, flag_field = entry[1], entry[2], entry[4]
            if flag_field == 1:
                print(f"File '{filename}' is deleted.")
                return
            fs.seek(start_offset)
            data = fs.read(length)
            ext = Path(filename).suffix.lower()
            if ext in [".png", ".jpg", ".jpeg", ".bmp", ".gif"]:
                temp_path = Path.cwd() / filename
                temp_path.write_bytes(data)
                output = climage.convert(temp_path, width=60)
                print(output)
                temp_path.unlink()
            else:
                try:
                    text = data.decode('utf-8')
                    print(text)
                except UnicodeDecodeError:
                    print("Warning: File is not valid text. Raw bytes output:")
                    print(data)


    def rem_file(self, fs_name: str, filename: str):
        fs_name = Path(fs_name)
        if not fs_name.exists():
            print(f"File system '{fs_name}' does not exist!")
            sys.exit(1)
        with fs_name.open("r+b") as fs:
            entry, entry_pos = self._find_file_entry(fs, filename)
            if not entry:
                print(f"File '{filename}' not found in filesystem.")
                return
            if entry[4] == 1:
                print(f"File '{filename}' is already deleted.")
                return
            new_entry_tuple = (entry[0], entry[1], entry[2], entry[3], 1, entry[5], entry[6], entry[7])
            fs.seek(entry_pos)
            fs.write(struct.pack(self.FILE_ENTRY_FORMAT, *new_entry_tuple))
            header = self._read_header(fs)
            header[4] -= 1
            header[12] += 1
            self._write_header(fs, header)
            print(f"Flagged '{filename}' as deleted.")


    def list_fs(self, fs_name: str):
        fs_name = Path(fs_name)
        if not fs_name.exists():
            print(f"File system '{fs_name}' does not exist!")
            sys.exit(1)
        with fs_name.open("rb") as fs:
            header = self._read_header(fs)
            table_offset = header[8]
            print(f"Files in filesystem '{fs_name}':\n")
            fs.seek(table_offset)
            found_any = False
            for _ in range(self.MAX_FILES):
                entry_data = fs.read(self.FILE_ENTRY_SIZE)
                if entry_data == b'\x00' * self.FILE_ENTRY_SIZE:
                    continue
                (name_bytes, _, length, _, flag_field, _, created_ts, _) = struct.unpack(self.FILE_ENTRY_FORMAT,
                                                                                         entry_data)
                if flag_field == 1:
                    continue
                found_any = True
                name_str = name_bytes.split(b"\x00")[0].decode(errors='ignore')
                try:
                    timestr = time.strftime("%Y-%m-%d %H:%M:%S", time.localtime(created_ts))
                except (OSError, ValueError):
                    timestr = "invalid timestamp"
                print(f"- Name: {name_str:<32} | Size: {length:<8} bytes | Created: {timestr}")
            if not found_any:
                print("(no active files)")


    def defrag_fs(self, fs_name: str):
        fs_name = Path(fs_name)
        if not fs_name.exists():
            print(f"File system '{fs_name}' does not exist!")
            sys.exit(1)
        valid_files = []
        with fs_name.open("rb") as fs:
            header = self._read_header(fs)
            table_offset = header[8]
            data_start_offset = header[9]
            original_next_free_offset = header[10]
            files_to_remove = header[12]
            fs.seek(table_offset)
            for _ in range(self.MAX_FILES):
                entry_data = fs.read(self.FILE_ENTRY_SIZE)
                if not entry_data or entry_data == b'\x00' * self.FILE_ENTRY_SIZE:
                    continue
                entry = struct.unpack(self.FILE_ENTRY_FORMAT, entry_data)
                if entry[4] == 0 and not entry[0].startswith(b'\x00'):
                    current_pos = fs.tell()
                    fs.seek(entry[1])
                    file_data = fs.read(entry[2])
                    fs.seek(current_pos)
                    valid_files.append({"entry": entry, "data": file_data})
        with fs_name.open("r+b") as fs:
            current_data_offset = data_start_offset
            new_file_table = b''
            for file_info in valid_files:
                original_entry = file_info['entry']
                data = file_info['data']
                length = original_entry[2]
                fs.seek(current_data_offset)
                fs.write(data)
                pad_len = (64 - (length % 64)) % 64
                if pad_len > 0:
                    fs.write(b'\x00' * pad_len)
                new_entry_tuple = (
                    original_entry[0], current_data_offset, length,
                    original_entry[3], original_entry[4], original_entry[5],
                    original_entry[6], original_entry[7]
                )
                new_file_table += struct.pack(self.FILE_ENTRY_FORMAT, *new_entry_tuple)
                current_data_offset += length + pad_len
            fs.seek(table_offset)
            fs.write(new_file_table)
            remaining_table_space = (self.MAX_FILES - len(valid_files)) * self.FILE_ENTRY_SIZE
            if remaining_table_space > 0:
                fs.write(b'\x00' * remaining_table_space)
            header = self._read_header(fs)
            header[4] = len(valid_files)
            header[10] = current_data_offset
            header[12] = 0
            self._write_header(fs, header)
            fs.truncate(current_data_offset)
            bytes_freed = original_next_free_offset - current_data_offset
            print(f"Defragmentation complete for '{fs_name}':")
            print(f"- Files removed: {files_to_remove}")
            print(f"- Files remaining: {len(valid_files)}")
            print(f"- Bytes freed: {bytes_freed}")
            print(f"- New next free offset: {current_data_offset}")


if __name__ == '__main__':
    if len(sys.argv) < 3:
        print("Usage:")
        print("  mkfs <file>")
        print("  gifs <file>")
        print("  lsfs <file>")
        print("  addfs <filesystem> <file_to_add>")
        print("  getfs <filesystem> <filename_to_get>")
        print("  catfs <filesystem> <filename_to_cat>")
        print("  rmfs <filesystem> <filename_to_remove>")
        print("  dfrgfs <filesystem>")
        sys.exit(1)

    cmd = sys.argv[1]
    fs_manager = ZestFileSystem(sys.argv[2])

    if cmd == "mkfs":
        fs_manager.create_new_fs(sys.argv[2])
    elif cmd == "gifs":
        fs_manager.get_info_fs(sys.argv[2])
    elif cmd == "addfs":
        if len(sys.argv) < 4:
            print("Usage: addfs <filesystem> <filepath>")
            sys.exit(1)
        fs_manager.add_file(sys.argv[2], sys.argv[3])
    elif cmd == "catfs":
        if len(sys.argv) < 4:
            print("Usage: catfs <filesystem> <filename>")
            sys.exit(1)
        fs_manager.cat_file(sys.argv[2], sys.argv[3])
    elif cmd == "getfs":
        if len(sys.argv) < 4:
            print("Usage: getfs <filesystem> <filename>")
            sys.exit(1)
        fs_manager.get_file(sys.argv[2], sys.argv[3])
    elif cmd == "rmfs":
        if len(sys.argv) < 4:
            print("Usage: rmfs <filesystem> <filename>")
            sys.exit(1)
        fs_manager.rem_file(sys.argv[2], sys.argv[3])
    elif cmd == "lsfs":
        fs_manager.list_fs(sys.argv[2])
    elif cmd == "dfrgfs":
        if len(sys.argv) < 3:
            print("Usage: dfrgfs <filesystem>")
            sys.exit(1)
        fs_manager.defrag_fs(sys.argv[2])
    else:
        print(f"Unknown command: '{cmd}'")
        sys.exit(1)

File: HS25_SoCo_group_040-a3/zvfs.java
import java.io.IOException;
import java.nio.ByteBuffer;
import java.nio.ByteOrder;
import java.nio.channels.FileChannel;
import java.nio.charset.StandardCharsets;
import java.nio.file.Files;
import java.nio.file.OpenOption;
import java.nio.file.Path;
import java.nio.file.Paths;
import java.nio.file.StandardCopyOption;
import java.nio.file.StandardOpenOption;
import java.util.ArrayList;
import java.util.Date;

public class zvfs { 
   public zvfs() {
   }

   static class Header {
      byte[] magic;
      byte version;
      byte flags;
      short reserved0;
      short fileCount;
      short fileCapacity;
      short fileEntrySize;
      short reserved1;
      int fileTableOffset;
      int dataStartOffset;
      int nextFreeOffset;
      int freeEntryOffset;
      short deletedFiles;
      byte[] reserved2;
   }

   static Header readHeader(FileChannel channel) throws IOException {
      System.out.println("reading header...");
      int headerSize = 64;
      ByteBuffer buffer = ByteBuffer.allocate(headerSize);
      buffer.order(ByteOrder.LITTLE_ENDIAN);
      channel.position(0L);

      int bytesRead = channel.read(buffer);
      if (bytesRead != headerSize) {
         throw new IOException("Could not read");
      }

      buffer.flip();
      Header header = new Header();
      header.magic = new byte[8];
      buffer.get(header.magic);
      header.version = buffer.get();
      header.flags = buffer.get();
      header.reserved0 = buffer.getShort();
      header.fileCount = buffer.getShort();
      header.fileCapacity = buffer.getShort();
      header.fileEntrySize = buffer.getShort();
      header.reserved1 = buffer.getShort();
      header.fileTableOffset = buffer.getInt();
      header.dataStartOffset = buffer.getInt();
      header.nextFreeOffset = buffer.getInt();
      header.freeEntryOffset = buffer.getInt();
      header.deletedFiles = buffer.getShort();
      header.reserved2 = new byte[26];
      buffer.get(header.reserved2);
      return header;
   }

   static void writeHeader(FileChannel channel, Header header) throws IOException {
      System.out.println("writing header...");
      int headerSize = 64;
      ByteBuffer buffer = ByteBuffer.allocate(headerSize);
      buffer.order(ByteOrder.LITTLE_ENDIAN);
      buffer.put(header.magic);
      buffer.put(header.version);
      buffer.put(header.flags);
      buffer.putShort(header.reserved0);
      buffer.putShort(header.fileCount);
      buffer.putShort(header.fileCapacity);
      buffer.putShort(header.fileEntrySize);
      buffer.putShort(header.reserved1);
      buffer.putInt(header.fileTableOffset);
      buffer.putInt(header.dataStartOffset);
      buffer.putInt(header.nextFreeOffset);
      buffer.putInt(header.freeEntryOffset);
      buffer.putShort(header.deletedFiles);
      buffer.put(header.reserved2);
      buffer.flip();
      channel.position(0L);
      channel.write(buffer);
   }

   static void mkfs(String filesystemName) throws Exception {
      System.out.println("mkfs starting...");
      Path filePath = Paths.get(filesystemName);
      if (Files.exists(filePath)) {
         System.err.println("File with the name already exists!!!!!");
      } else {
         FileChannel channel = FileChannel.open(Paths.get(filesystemName),
                 StandardOpenOption.CREATE,
                 StandardOpenOption.READ,
                 StandardOpenOption.WRITE);

         try {
            byte headerSize = 64;
            byte fileEntrySize = 64;
            byte fileCapacity = 32;
            ByteBuffer headerBuffer = ByteBuffer.allocate(headerSize);
            headerBuffer.order(ByteOrder.LITTLE_ENDIAN);
            byte[] magicAscii = "ZVFSDSK1".getBytes(StandardCharsets.US_ASCII);
            byte[] magicBytes = new byte[8];
            System.arraycopy(magicAscii, 0, magicBytes, 0, 8);
            headerBuffer.put(magicBytes);
            headerBuffer.put((byte)1);
            headerBuffer.put((byte)0);
            headerBuffer.putShort((short)0);
            headerBuffer.putShort((short)0);
            headerBuffer.putShort((short)fileCapacity);
            headerBuffer.putShort((short)fileEntrySize);
            headerBuffer.putShort((short)0);
            headerBuffer.putInt(headerSize);
            int dataStartOffset = headerSize + fileCapacity * fileEntrySize;
            headerBuffer.putInt(dataStartOffset);
            headerBuffer.putInt(dataStartOffset);
            headerBuffer.putInt(0);
            headerBuffer.putShort((short)0);
            headerBuffer.put(new byte[26]);
            headerBuffer.flip();
            channel.position(0L);
            channel.write(headerBuffer);
            ByteBuffer entryRegionBuffer = ByteBuffer.allocate(fileCapacity * fileEntrySize);
            entryRegionBuffer.order(ByteOrder.LITTLE_ENDIAN);

            while(true) {
               if (!entryRegionBuffer.hasRemaining()) {
                  entryRegionBuffer.flip();
                  channel.position((long)headerSize);
                  channel.write(entryRegionBuffer);
                  break;
               }

               entryRegionBuffer.put((byte)0);
            }
         } catch (Exception e) {
            throw new Exception("---------sth wrong: ", e);
         }
      }
   }

   static FileEntry readEntry(FileChannel channel, int entryIndex, int tableOffset, int entrySize) throws IOException {
      long position = (long)tableOffset + (long)entryIndex * (long)entrySize;
      ByteBuffer buffer = ByteBuffer.allocate(entrySize);
      buffer.order(ByteOrder.LITTLE_ENDIAN);
      channel.position(position);
      int bytesRead = channel.read(buffer);
      if (bytesRead != entrySize) {
         return null;
      } else {
         buffer.flip();
         byte[] nameBytes = new byte[32];
         buffer.get(nameBytes);
         int start = buffer.getInt();
         int length = buffer.getInt();
         byte type = buffer.get();
         byte deletedFlag = buffer.get();
         buffer.getShort();
         long created = buffer.getLong();
         byte[] reservedBytes = new byte[12];
         buffer.get(reservedBytes);
         boolean allZero = true;

         for(int i = 0; i < nameBytes.length; ++i) {
            if (nameBytes[i] != 0) {
               allZero = false;
               break;
            }
         }

         String fileName = allZero ? "" : (new String(nameBytes, StandardCharsets.UTF_8)).split("\u0000", 2)[0];
         FileEntry entry = new FileEntry();
         entry.name = fileName;
         entry.start = start;
         entry.length = length;
         entry.type = type;
         entry.deleted = deletedFlag == 1;
         entry.created = created;
         return entry;
      }
   }

   static void writeEntry(FileChannel channel, int entryIndex, int tableOffset, int entrySize, FileEntry entry) throws IOException {
      long position = (long)tableOffset + (long)entryIndex * (long)entrySize;
      ByteBuffer buffer = ByteBuffer.allocate(entrySize);
      buffer.order(ByteOrder.LITTLE_ENDIAN);
      byte[] nameBytes = new byte[32];
      byte[] encodedName = entry.name == null ? new byte[0] : entry.name.getBytes(StandardCharsets.UTF_8);
      int copyLength = Math.min(31, encodedName.length);
      System.arraycopy(encodedName, 0, nameBytes, 0, copyLength);
      buffer.put(nameBytes);
      buffer.putInt(entry.start);
      buffer.putInt(entry.length);
      buffer.put(entry.type);
      buffer.put((byte)(entry.deleted ? 1 : 0));
      buffer.putShort((short)0);
      buffer.putLong(entry.created);
      buffer.put(new byte[12]);
      buffer.flip();
      channel.position(position);
      channel.write(buffer);
   }

   static void gifs(String filesystemName) throws IOException {
      Path filePath = Paths.get(filesystemName);
      if (!Files.exists(filePath)) {
         System.err.println("Does not exist!!!!");
      } else {
         FileChannel channel = FileChannel.open(filePath, StandardOpenOption.READ);

         try {
            Header header = readHeader(channel);
            // String magic = new String(header.magic, StandardCharsets.US_ASCII);
            long size = Files.size(filePath);
            int remaining = header.fileCapacity - header.fileCount - header.deletedFiles;
            System.out.println("*****GIFS*****");
            System.out.println("name: " + filesystemName);
            System.out.println("number of files present: " + (header.fileCount - header.deletedFiles));
            System.out.println("remaining free entries: " + remaining);
            System.out.println("Deleted files - marked: " + header.deletedFiles);
            System.out.println("Total size: " + size);
         } finally {
            if (channel != null) {
               channel.close();
            }
         }
      }
   }

   static void lsfs(String filesystemName) throws IOException {
      Path filePath = Paths.get(filesystemName);
      if (!Files.exists(filePath)) {
         System.err.println("Does not exist!!!!");
      } else {
         FileChannel channel = FileChannel.open(filePath, StandardOpenOption.READ);

         try {
            Header header = readHeader(channel);
            for(int i = 0; i < header.fileCapacity; ++i) {
               FileEntry entry = readEntry(channel, i, header.fileTableOffset, header.fileEntrySize);
               if (entry != null && entry.name != null && !entry.name.isEmpty() && !entry.deleted) {
                  Date date = new Date(entry.created * 1000L);
                  System.out.println("*****LSFS*****");
                  System.out.println("name: " + entry.name);
                  System.out.println("size: " + entry.length);
                  System.out.println("created: " + date);
               }
            }
         } finally {
            if (channel != null) {
               channel.close();
            }
         }
      }
   }

   static void rmfs(String filesystemName, String fileName) throws IOException {
      if (fileName != null && !fileName.isEmpty()) {
         Path filePath = Paths.get(filesystemName);
         if (!Files.exists(filePath)) {
            System.err.println("Does not exist");
         } else {
            FileChannel channel = FileChannel.open(Paths.get(filesystemName),
                    StandardOpenOption.CREATE,
                    StandardOpenOption.READ,
                    StandardOpenOption.WRITE);

            try {
               Header header = readHeader(channel);
               int targetIndex = -1;
               FileEntry targetEntry = null;

               for(int i = 0; i < header.fileCapacity; ++i) {
                  FileEntry candidate = readEntry(channel, i, header.fileTableOffset, header.fileEntrySize);
                  if (candidate != null && candidate.name != null && !candidate.name.isEmpty() && !candidate.deleted && candidate.name.equals(fileName)) {
                     targetIndex = i;
                     targetEntry = candidate;
                     break;
                  }
               }

               if (targetIndex != -1 && targetEntry != null) {
                  targetEntry.deleted = true;
                  writeEntry(channel, targetIndex, header.fileTableOffset, header.fileEntrySize, targetEntry);
                  header.fileCount--;
                  header.deletedFiles++;
                  header.flags = 0;
                  writeHeader(channel, header);
               } else {
                  System.err.println("FFNF!!");
               }
            } finally {
               if (channel != null) {
                  channel.close();
               }
            }
         }
      } else {
         System.err.println("Invvalid");
      }
   }

   static void addfs(String filesystemName, String sourceName) throws IOException {
      if (sourceName != null && !sourceName.isEmpty()) {
         Path filePath = Paths.get(filesystemName);
         if (!Files.exists(filePath)) {
            System.err.println("does not exist");
         } else {
            Path sourceFilePath = Paths.get(sourceName);
            if (!Files.exists(sourceFilePath)) {
               System.err.println("Source does not exist!!");
            } else {
               FileChannel channel = FileChannel.open(Paths.get(filesystemName),
                       StandardOpenOption.CREATE,
                       StandardOpenOption.READ,
                       StandardOpenOption.WRITE);

               try {
                  Header header = readHeader(channel);
                  if (header.fileCount + header.deletedFiles >= header.fileCapacity) {
                     System.err.println("No free slots");
                     return;
                  }

                  int freeIndex = -1;

                  for(int i = 0; i < header.fileCapacity; ++i) {
                     FileEntry candidate = readEntry(channel, i, header.fileTableOffset, header.fileEntrySize);
                     if (candidate == null || candidate.name == null || candidate.name.isEmpty() || candidate.deleted) {
                        freeIndex = i;
                        break;
                     }
                  }

                  if (freeIndex == -1) {
                     System.err.println("No free entry slots");
                     return;
                  }

                  byte[] fileData = Files.readAllBytes(sourceFilePath);
                  int alignedStart = (header.nextFreeOffset + 63) / 64 * 64;
                  ByteBuffer dataBuffer = ByteBuffer.wrap(fileData);
                  channel.position((long)alignedStart);

                  while(dataBuffer.hasRemaining()) {
                     channel.write(dataBuffer);
                  }

                  FileEntry newEntry = new FileEntry();
                  newEntry.name = sourceFilePath.getFileName().toString();
                  newEntry.start = alignedStart;
                  newEntry.length = fileData.length;
                  newEntry.type = 0;
                  newEntry.deleted = false;
                  newEntry.created = System.currentTimeMillis() / 1000L;
                  writeEntry(channel, freeIndex, header.fileTableOffset, header.fileEntrySize, newEntry);
                  header.fileCount++;
                  header.nextFreeOffset = alignedStart + fileData.length;
                  header.flags = (byte)((header.fileCount + header.deletedFiles >= header.fileCapacity) ? 1 : 0);
                  writeHeader(channel, header);
               } finally {
                  if (channel != null) {
                     channel.close();
                  }
               }
            }
         }
      } else {
         System.err.println("Invalid");
      }
   }

   static void getfs(String filesystemName, String fileName) throws IOException {
      if (fileName != null && !fileName.isEmpty()) {
         Path filePath = Paths.get(filesystemName);
         if (!Files.exists(filePath)) {
            System.err.println("-does not exist");
         } else {
            FileChannel channel = FileChannel.open(filePath, StandardOpenOption.READ);

            try {
               Header header = readHeader(channel);
               FileEntry target = null;

               for(int i = 0; i < header.fileCount; ++i) {
                  FileEntry candidate = readEntry(channel, i, header.fileTableOffset, header.fileEntrySize);
                  if (candidate != null && candidate.name != null && !candidate.name.isEmpty() && !candidate.deleted && candidate.name.equals(fileName)) {
                     target = candidate;
                     break;
                  }
               }

               if (target != null) {
                  ByteBuffer dataBuffer = ByteBuffer.allocate(target.length);
                  channel.position((long)target.start);
                  int bytesRead = channel.read(dataBuffer);
                  if (bytesRead != target.length) {
                     System.err.println("Could not read file!!");
                     return;
                  }

                  dataBuffer.flip();
                  byte[] fileBytes = new byte[target.length];
                  dataBuffer.get(fileBytes);
                  Files.write(Paths.get(target.name), fileBytes, new OpenOption[0]);
               } else {
                  System.err.println("Not found!!!!");
               }
            } finally {
               if (channel != null) {
                  channel.close();
               }
            }
         }
      } else {
         System.err.println("Invalid--");
      }
   }

   static void catfs(String fsPath, String fileName) throws IOException {
      if (fileName == null || fileName.isEmpty()) {
         System.err.println("Invalid--");
         return;
      }

      Path fs = Paths.get(fsPath);
      if (!Files.exists(fs)) {
         System.err.println("--Does not exist!!");
         return;
      }

      try (FileChannel ch = FileChannel.open(fs, StandardOpenOption.READ)) {
         Header header = readHeader(ch);
         FileEntry target = null;
         for (int i = 0; i < header.fileCapacity; i++) {
            FileEntry e = readEntry(ch, i, header.fileTableOffset, header.fileEntrySize);
            if (e != null && e.name != null && !e.name.isEmpty() && !e.deleted && e.name.equals(fileName)) {
               target = e;
               break;
            }
         }

         if (target == null) {
            System.err.println("FNF!!");
            return;
         }

   ByteBuffer dataBuf = ByteBuffer.allocate(target.length);
   ch.position((long) target.start);
   int bytesRead = ch.read(dataBuf);
   if (bytesRead != target.length) {
      System.err.println("--DBG!");
      return;
   }
   dataBuf.flip();
   byte[] data = new byte[target.length];
   dataBuf.get(data);
   String content = new String(data, StandardCharsets.UTF_8);
   System.out.print(content);

      }
   }

   static void dfrgfs(String fsPath) throws Exception {
      /// DEFRAGMENTATION logic ---- MOSTLY WITH CHATGPT!, see readme
      Path src = Paths.get(fsPath);
      if (!Files.exists(src)) {
         System.err.println("does not exist");
         return;
      }

      try (FileChannel in = FileChannel.open(src, StandardOpenOption.READ)) {
         Header header = readHeader(in);
         if (header.deletedFiles == 0) {
            System.out.println("defragmented 0 files, freed 0 bytes");
            return;
         }

         ArrayList<FileEntry> active = new ArrayList<>();
         for (int i = 0; i < header.fileCapacity; i++) {
            FileEntry e = readEntry(in, i, header.fileTableOffset, header.fileEntrySize);
            if (e != null && e.name != null && !e.name.isEmpty() && !e.deleted) {
               active.add(e);
            }
         }

         long oldUsed = (long) header.nextFreeOffset - (long) header.dataStartOffset;

         String tmpName = fsPath + ".tmp";
         Path tmpPath = Paths.get(tmpName);
         Files.deleteIfExists(tmpPath);
         mkfs(tmpName);

         try (FileChannel out = FileChannel.open(Paths.get(tmpName),
                 StandardOpenOption.CREATE,
                 StandardOpenOption.READ,
                 StandardOpenOption.WRITE)) {

            Header tempHeader = readHeader(out);
            int writePos = tempHeader.dataStartOffset;
            int index = 0;

            for (FileEntry e : active) {
               int aligned = ((writePos + 63) / 64) * 64;
               ByteBuffer dataBuf = ByteBuffer.allocate(e.length);
               in.position((long) e.start);
               int read = in.read(dataBuf);
               if (read != e.length) {
                  System.err.println("Error reading file data during defrag");
                  return;
               }
               dataBuf.flip();
               out.position((long) aligned);
               while (dataBuf.hasRemaining()) {
                  out.write(dataBuf);
               }
               FileEntry ne = new FileEntry();
               ne.name = e.name;
               ne.start = aligned;
               ne.length = e.length;
               ne.type = e.type;
               ne.deleted = false;
               ne.created = e.created;
               writeEntry(out, index, tempHeader.fileTableOffset, tempHeader.fileEntrySize, ne);
               index++;
               writePos = aligned + e.length;
            }

            tempHeader.fileCount = (short) active.size();
            tempHeader.deletedFiles = 0;
            tempHeader.nextFreeOffset = ((writePos + 63) / 64) * 64;
            tempHeader.flags = (byte) ((tempHeader.fileCount == tempHeader.fileCapacity) ? 1 : 0);
            writeHeader(out, tempHeader);

            long newUsed = (long) tempHeader.nextFreeOffset - (long) tempHeader.dataStartOffset;
            long freed = oldUsed - newUsed;

            Files.move(tmpPath, src, StandardCopyOption.REPLACE_EXISTING);
            System.out.println("# " + header.deletedFiles + "defragmented files, empty " + freed + " bytes");
         }
      }
   }

   public static void main(String[] args) throws Exception {
      if (args.length < 1) {
         System.err.println("Needs args!!");
      } else {
         switch (args[0]) {
            case "mkfs":
               mkfs(args[1]);
               break;
            case "gifs":
               gifs(args[1]);
               break;
            case "lsfs":
               lsfs(args[1]);
               break;
            case "rmfs":
               rmfs(args[1], args[2]);
               break;
            case "addfs":
               addfs(args[1], args[2]);
               break;
            case "getfs":
               getfs(args[1], args[2]);
               break;
            case "catfs":
               catfs(args[1], args[2]);
               break;
            case "dfrgfs":
               dfrgfs(args[1]);
               break;
            default:
               System.err.println("Unknown command!!!");
         }

      }
   }

   static class FileEntry {
      String name;
      int start;
      int length;
      byte type;
      boolean deleted;
      long created;

      FileEntry() {
      }
   }
}
